pkg/kv/kvserver/logstore/bytes_tracker.go

```diff
@@ -0,0 +1,134 @@
+// licenses/APL.txt.
+
+package logstore
+
+import (
+	"github.com/cockroachdb/cockroach/pkg/settings"
+	"github.com/cockroachdb/cockroach/pkg/util/metric"
+)
+
+// MaxRaftEntriesSize is the "global" soft limit for the total size of raft log
+// entries pulled into memory simultaneously. Currently, this only includes the
+// entries pulled as part of the local state machine application flow.
+//
+// TODO(pav-kv): track/limit usage for other large memory consumers in raft:
+//   - the leader->follower append flows,
+//   - the raft/unstable data structure.
+var MaxRaftEntriesSize = settings.RegisterByteSizeSetting(
+	settings.SystemOnly,
+	"kv.raft.loaded_entries.max_size",
+	"maximum size of raft entries pulled into memory simultaneously (0 means no limit)",
+	0,
+)
+
+// The reservation sizes are optimized with the following constraints in mind:
+//
+//   - Per kvserver.defaultRaftSchedulerConcurrency, there are <= 128 raft
+//     scheduler workers.
+//   - Per bulk.IngestBatchSize and kvserverbase.MaxCommandSize, raft entries are
+//     up to 16 MB in size. Entries are typically small, but can reach this
+//     limit, e.g. if these are AddSSTable commands or large batch INSERTs.
+//     Typically, compression reduces the max size to < 8MB.
+const (
+	// minReserveSize is the granularity at which bytes are reserved from the
+	// SoftLimiter for small allocations (below smallReserveSize). Batching small
+	// allocations this way amortizes the runtime cost of SoftLimiter which is
+	// typically shared by many goroutines.
+	minReserveSize = 256 << 10
+	// smallReserveSize is the threshold below which the size is considered
+	// "small", and causes reserving in multiples of minReserveSize at a time.
+	smallReserveSize = 4 << 20
+)
+
+// SoftLimiter is a byte size limiter with weakened guarantees. It tracks the
+// global usage in a metric, and gives the user a hint when the usage has
+// reached the soft limit.
+//
+// When at the limit, the user is supposed to back off acquiring resources, but
+// might choose to do some work regardless if there is a risk of stalling, and
+// the total over-the-limit overhead is bounded (for instance, there is a
+// limited amount of workers, and the active resource per worker is bounded).
+//
+// TODO(pav-kv): integrate with mon.BytesMonitor.
+type SoftLimiter struct {
+	Metric *metric.Gauge    // the "global" usage metric
+	SV     *settings.Values // contains the soft limit, can change dynamically
+}
+
+// acquire accepts the given number of bytes for tracking.
+func (s *SoftLimiter) acquire(x int64) {
+	s.Metric.Inc(x)
+}
+
+// withinLimit returns true iff the bytes usage hasn't reached the soft limit.
+func (s *SoftLimiter) withinLimit() bool {
+	limit := MaxRaftEntriesSize.Get(s.SV)
+	return limit <= 0 || s.Metric.Value() < limit
+}
+
+// release removes the given number of bytes from tracking.
+func (s *SoftLimiter) release(x int64) {
+	s.Metric.Dec(x)
+}
+
+type BytesAccount struct {
```

Comment:
Need to add some unit-tests for these new types. For now only tested manually, and observed that `kv.raft.loaded_entries.max_size` helps reducing peak memory footprint on my test from 10GB to 6GB. Also observed that the bytes in the tracker don't leak.